ujt/state.py

```diff
@@ -258,3 +245,225 @@ def set_virtual_node_collapsed_state(virtual_node_name: str, collapsed: bool):
+            save_changes = True
+        if node_name in proto_map:  # type: ignore
+            proto_map[  # type: ignore
+                node_name].override_status = override_status  # type: ignore
+            if save_changes:
+                setters[idx](proto_map)
+
+    override_status_map = cache.get("override_status_map")
+    if override_status == Status.STATUS_UNSPECIFIED:
+        try:
+            del override_status_map[node_name]
+        except KeyError:  # the node didn't have a comment before
+            pass
+    else:
+        override_status_map[node_name] = override_status
+    cache.set("override_status_map", override_status_map)
+
+
+#@cache.memoize()
+def get_node_to_user_journey_map() -> Dict[str, List[UserJourney]]:
+    # map the node name to user journey names that pass through the node
+    # should this be in state? it's memoized but doesn't really affect state
+    # however, it's kind of similar to the other maps we expose in state.py,
+    # only this one is dynamically generated once...
+    node_name_message_map, client_name_message_map = get_message_maps()
+    output_map: DefaultDict[
+        str,
+        List[UserJourney]] = defaultdict(
+            list
+        )  # we would prefer to use a set here, but protobufs are not hashable
+    for client in client_name_message_map.values():
+        for user_journey in client.user_journeys:
+            node_frontier = deque(
+                [
+                    dependency.target_name
+                    for dependency in user_journey.dependencies
+                ])
+
+            while node_frontier:
+                current_node_name = node_frontier.popleft()
+                if user_journey not in output_map[
+                        current_node_name]:  # since we don't use a set, we check if it exists in the list already
+                    output_map[current_node_name].append(user_journey)
+
+                # add all the node's parents
+                parent_name = node_name_message_map[
+                    current_node_name].parent_name
+                while parent_name != "":
+                    if user_journey not in output_map[parent_name]:
+                        output_map[parent_name].append(user_journey)
+                    parent_name = node_name_message_map[parent_name].parent_name
+
+                for dependency in node_name_message_map[
+                        current_node_name].dependencies:
+                    node_frontier.append(dependency.target_name)
+
+    return output_map
+
+
+def get_tag_list() -> List[str]:
+    """ Return the list of tags.
+
+    We use a list since the order matters when using pattern matching callbacks to remove tags.
+
+    """
+    return cache.get("tag_list")
+
+
+def set_tag_list(tag_list):
+    return cache.set("tag_list", tag_list)
+
+
+def create_tag(new_tag):
+    tag_list = get_tag_list()
+    if " " in new_tag:  # where should validation be done? in the callback or here?
```

Comment:
i think here makes sense

Reply:
I think it's reasonable to do it here in a "preserve data integrity" context but we need to do it in callbacks if we want to display some error to the user. It doesn't hurt to have it in both places, right?